setuptools/command/easy_install.py

```diff
@@ -2055,19 +2059,18 @@ def _sys_executable(cls):
         return os.environ.get('__PYVENV_LAUNCHER__', _default)
 
     @classmethod
-    def from_param(cls, param):
+    def from_param(cls, param: Self | str | Iterable[str] | None) -> Self:
         """
         Construct a CommandSpec from a parameter to build_scripts, which may
         be None.
         """
         if isinstance(param, cls):
             return param
-        if isinstance(param, list):
+        if isinstance(param, str):
+            return cls.from_string(param)
+        if isinstance(param, Iterable):
             return cls(param)
-        if param is None:
-            return cls.from_environment()
-        # otherwise, assume it's a string.
-        return cls.from_string(param)
+        return cls.from_environment()
```

Comment:
If this gracious fallback is undesired, I can explicitly raise instead.

Reply:
Let's raise it explicitly because before `cls.from_environment()` was only used in the case of `None`. I suspect that the previous fallback to `cls.from_string` would cause a `TypeError` if the user passed non-sensical args.

We don't want to change that behaviour

Reply:
Raising an error instead of a fallback makes sense since the param unexpectedly goes unused.

It was an `AttributeError` because non-string first argument to `shlex` is assumed to be a stream.
```py
  File "C:\Program Files\Python39\lib\shlex.py", line 140, in read_token
    nextchar = self.instream.read(1)
AttributeError: 'int' object has no attribute 'read'
```
At the `from_param` level it should really be a TypeError, but that's technically a non-backwards compatible change.

I raise an `AttributeError`, maybe can be changed for a `TypeError` next major version (or now if you don't mind the error type change)

Reply:
Thanks @Avasam, I am happy with either.

I don't think that changing to `TypeError` should be a big impact because it looks like a very edge-y case.
Probably the important part is to raise the exception, not the exact exception class. So I am OK with risking it.

But equally happy to take the conservative approach, given how susceptible `setuptools` is to Hyrum's Law...

Reply:
How about this: Keep it to `AttributeError` for this PR, I'll open a follow-up that changes to `TypeError` that you can merge before the next major update. That should cover our bases and improve the communicated error.

Reply:
Thanks @Avasam.